pkg/cedar/core/client.go

```diff
@@ -0,0 +1,91 @@
+package cedarcore
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	ld "gopkg.in/launchdarkly/go-server-sdk.v5"
+
+	"github.com/go-openapi/runtime"
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	"go.uber.org/zap"
+
+	"github.com/cmsgov/easi-app/pkg/appcontext"
+	apiclient "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client"
+	apisystems "github.com/cmsgov/easi-app/pkg/cedar/core/gen/client/system"
+	coremodels "github.com/cmsgov/easi-app/pkg/cedar/core/gen/models"
+	"github.com/cmsgov/easi-app/pkg/flags"
+)
+
+const (
+	cedarCoreEnabledKey     = "cedarCoreEnabled"
+	cedarCoreEnabledDefault = false
+)
+
+// NewClient builds the type that holds a connection to the CEDAR Core API
+func NewClient(cedarHost string, cedarAPIKey string, ldClient *ld.LDClient) *Client {
+	fnEmit := func(ctx context.Context) bool {
+		lduser := flags.Principal(ctx)
+		result, err := ldClient.BoolVariation(cedarCoreEnabledKey, lduser, cedarCoreEnabledDefault)
+		if err != nil {
+			appcontext.ZLogger(ctx).Info(
+				"problem evaluating feature flag",
+				zap.Error(err),
+				zap.String("flagName", cedarCoreEnabledKey),
+				zap.Bool("flagDefault", cedarCoreEnabledDefault),
+				zap.Bool("flagResult", result),
+			)
+		}
+		return result
+	}
+
+	hc := http.DefaultClient
+
+	return &Client{
+		cedarCoreEnabled: fnEmit,
+		auth: httptransport.APIKeyAuth(
+			"x-Gateway-APIKey",
+			"header",
+			cedarAPIKey,
+		),
+		sdk: apiclient.New(
+			httptransport.New(
+				cedarHost,
+				apiclient.DefaultBasePath,
+				apiclient.DefaultSchemes,
+			),
+			strfmt.Default,
+		),
+		hc: hc,
+	}
+}
+
+// Client represents a connection to the CEDAR Core API
+type Client struct {
+	cedarCoreEnabled func(context.Context) bool
+	auth             runtime.ClientAuthInfoWriter
+	sdk              *apiclient.CEDARCoreAPI
+	hc               *http.Client
+}
+
+// GetSystemSummary makes a GET call to the /system/summary endpoint
+func (c *Client) GetSystemSummary(ctx context.Context) (coremodels.SystemSummaryResponse, error) {
```

Comment:
Comparing this to `pkg/cedar/cedarldap/translated_client.go`'s `FetchUserInfo()` method, that returns a model we define in `pkg/models`, rather than an autogenerated model from the Swagger schema. I feel like we should do something similar here; part of the reason for having our own `Client` setup is to provide insulation between the CEDAR API types and our own domain types.

Reply:
I just pushed an update that shows one way of doing this, with another method in there (but commented out)

There's some pros and cons to each approach that I've been able to think up

Marshal and Unmarshal (commented solution):
- Doesn't require you to loop through the result array and name out each property
- Relies on the tags in the model to pull properties out
- If a property name changed (e.g. `description` becomes `desc`, it would _NOT_ result in an error, and would silently make `description` be `""` in future structs

Loop:
- More verbose, manually creating the structs for the response
- If a property name changed or was removed, it'd result in a compiler error and would be more easily caught.

@DylanSpOddball @downeyn-cms Any thoughts on these?

Reply:
As discussed, I like the loop for the compile-time safety it is.

Reply:
Implemented the looping method - should be all set!